src/circuit/gadget/ecc/chip.rs

```diff
@@ -0,0 +1,473 @@
+use super::EccInstructions;
+use crate::constants;
+use ff::Field;
+use halo2::{
+    arithmetic::CurveAffine,
+    circuit::{Cell, Chip, Layouter},
+    plonk::{Advice, Column, ConstraintSystem, Error, Fixed, Permutation, Selector},
+};
+
+mod add;
+mod add_incomplete;
+#[cfg(test)]
+mod double;
+mod load;
+mod mul;
+mod mul_fixed;
+mod util;
+mod witness_point;
+mod witness_scalar_fixed;
+
+pub use load::*;
+
+/// A structure containing a cell and its assigned value.
+#[derive(Clone, Debug)]
+pub struct CellValue<T> {
+    /// The cell of this `CellValue`
+    pub cell: Cell,
+    /// The value assigned to this `CellValue`
+    pub value: Option<T>,
+}
+
+impl<T> CellValue<T> {
+    /// Construct a `CellValue`.
+    pub fn new(cell: Cell, value: Option<T>) -> Self {
+        CellValue { cell, value }
+    }
+}
+
+/// A curve point represented in affine (x, y) coordinates. Each coordinate is
+/// assigned to a cell.
+#[derive(Clone, Debug)]
+pub struct EccPoint<C: CurveAffine> {
+    /// x-coordinate
+    pub x: CellValue<C::Base>,
+    /// y-coordinate
+    pub y: CellValue<C::Base>,
+}
+
+impl<C: CurveAffine> EccPoint<C> {
+    /// Returns the value of this curve point, if known.
+    pub fn point(&self) -> Option<C> {
+        match (self.x.value, self.y.value) {
+            (Some(x), Some(y)) => {
+                if x == C::Base::zero() && y == C::Base::zero() {
+                    Some(C::identity())
+                } else {
+                    Some(C::from_xy(x, y).unwrap())
+                }
+            }
+            _ => None,
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+/// For each Orchard fixed base, we precompute:
+/// * coefficients for x-coordinate interpolation polynomials, and
+/// * z-values such that y + z = u^2 some square while -y + z is non-square.
+pub struct EccLoaded<C: CurveAffine> {
+    commit_ivk_r: OrchardFixedBase<C>,
+    note_commit_r: OrchardFixedBase<C>,
+    nullifier_k: OrchardFixedBase<C>,
+    value_commit_r: OrchardFixedBase<C>,
+    value_commit_v: OrchardFixedBaseShort<C>,
+}
+
+/// Configuration for the ECC chip
+#[derive(Clone, Debug)]
+#[allow(non_snake_case)]
+pub struct EccConfig {
+    /// Advice column for scalar decomposition into bits
+    pub bits: Column<Advice>,
+    /// Holds a point (x_p, y_p)
+    pub P: (Column<Advice>, Column<Advice>),
```

Comment:
We also use these columns (or at least the first one) to store scalars right?

Reply:
I'm actually inclined to pass in `advices: [Column<Advice>; 10]` instead of naming any columns or "extras" here. The `bits`, `P`, and `lambda` columns are renamed in some operations anyway.

Reply:
I've made it so that the top-level `EccConfig` only knows about 10 generic advice columns. These columns are renamed by individual components (e.g. `AddConfig`, `MulConfig`). The mapping for each individual config is captured in its `impl From<EccConfig>`.